src/extensions/Statiq.Html/ValidateLinks.cs

```diff
@@ -330,16 +330,28 @@ private static async Task<bool> ValidateAbsoluteLinkAsync(Uri uri, HttpMethod me
             }
         }
 
-        private static void AddOrUpdateLink(string link, IElement element, string source, ConcurrentDictionary<string, ConcurrentBag<(string source, string outerHtml)>> links)
+        private static void AddOrUpdateLink(string link, IElement element, IDocument source, ConcurrentDictionary<string, ConcurrentBag<(IDocument source, string outerHtml)>> links)
         {
             if (string.IsNullOrEmpty(link))
             {
                 return;
             }
 
+            // Attempt to parse the URI
+            if (Uri.TryCreate(link, UriKind.RelativeOrAbsolute, out Uri uri))
+            {
+                // If this is a relative Uri, adjust it relative to the source document.
+                // Adjustment for double-slash link prefix which means use http:// or https:// depending on current protocol
+                // The Uri class treats these as relative, but they're really absolute
+                if (!uri.IsAbsoluteUri && !uri.ToString().StartsWith("//") && !source.Destination.IsNull)
+                {
+                    link = source.Destination.Parent.Combine(new NormalizedPath(uri.ToString())).FullPath;
+                }
+            }
```

Comment:
I would really like to avoid duplicating this, as it's also present in the Link validation. We can't just go and throw the link as key, because link `../foo.html` are actually different if one comes from `input.html` and another from `input/input.html`.

Reply:
So the problem is we need a key that's unique to the original link value _and_ the source document path? That way `../foo` from `a/b.html` and `x/y/z.html` don't have the same key?

What about a tuple key with components for both the original link test and the parent folder of the source document (that way source documents from the same folder will still cache the same key)?